src/eventSourcing/EventSourcing.js

```diff
@@ -0,0 +1,188 @@
+const Event = require('./Event')
+const Workflow = require('../workflow')
+
+// Important: the event sourcing interface relies on the events being
+// sorted by createdAt in ascending order!
+// (otherwise every function that relies on the order would need to sort them
+// which can result in a lot of redundant sorting)
+
+const ACTIVATION_EVENTS = {
+  [Workflow.EventListener.ELEMENT_TYPE]: Event.TYPES.EVENT_LISTENER_ACTIVATED,
+  [Workflow.Flow.ELEMENT_TYPE]: Event.TYPES.FLOW_SIGNALED,
+  [Workflow.Gateway.ELEMENT_TYPE]: Event.TYPES.GATEWAY_ACTIVATED,
+  [Workflow.EventDispatcher.ELEMENT_TYPE]: Event.TYPES.DISPATCHED_EVENT,
+  [Workflow.Phase.ELEMENT_TYPE]: Event.TYPES.PHASE_STARTED,
+}
+
+const ELEMENT_TYPE_TO_ID_KEY = {
+  [Workflow.EventListener.ELEMENT_TYPE]: 'eventListenerId',
+  [Workflow.Flow.ELEMENT_TYPE]: 'flowId',
+  [Workflow.Gateway.ELEMENT_TYPE]: 'gatewayId',
+  [Workflow.EventDispatcher.ELEMENT_TYPE]: 'eventDispatcherId',
+  [Workflow.Phase.ELEMENT_TYPE]: 'phaseId',
+}
+
+const getEventsByType = (events, targetType) => (
+  events.filter(({ type }) => type === targetType)
+)
+
+const getStartedEvent = events => (
+  // This implementation could be dangerous - we are assuming the first "EVENT_RECEIVED"
+  // event is the start event; This way we don't have to look up anything on workflow definition;
+  // reality might show cases where this is not true even though I
+  // coudn't come up with anything in my fantasy; then I suggest adding a "WORKFLOW_STARTED"
+  // event type that is additionally dispatched by the start event listener
+  events.find(({ type }) => (
+    type === Event.TYPES.EVENT_RECEIVED
+  ))
+)
+
+const getEndingEvent = events => (
+  // a workflow can either be terminated (forcefully stopped) or finished (when running through)
+  events.find(({ type }) => type === Event.TYPES.WORKFLOW_TERMINATED
+    || type === Event.TYPES.WORKFLOW_FINISHED)
+)
+
+const getWorkflowDuration = events => {
```

Comment:
In theory, not only the workflow duration, but also the Workflow state can be completely derived from the events. We still want to keep our `state`, `finishedAt` and `startedAt` on the process though, because it makes querying the database much easier. Further, when we fetch a list of processes, without those keys we would need to fetch the events of every single process, which is a bot too much I think.